### PR TITLE
feat: add client id parameter into bw query

### PR DESCRIPTION
**Pull request by limitofzero**
# Summary

Attach a browser-scoped client_id to every balances-watcher request. Backend indexes sessions by `(chainId, owner, clientId)`. Backend doesn't support session-id yet, the frontend pr should be merged first otherwise it breaks the session. 

# To Test

1. Open the swap UI on any EVM chain with a connected wallet

- [ ] `POST /sessions/*` request in DevTools Network tab has an `X-Client-Id` header
- [ ] `/sse/*/balances/*` request URL includes `?client_id=<uuid>` query parameter
- [ ] Both values are the same UUID
- [ ] Refresh the page — the same `client_id` value is used (persisted)

2. In DevTools Application tab → Local Storage

- [ ] Key `balances-watcher-client-id` exists
- [ ] Value is JSON `{"id": "<uuid>", "createdAt": <timestamp>}`

3. TTL rotation (manual)

- [ ] Overwrite `createdAt` in localStorage to a value older than 24h, refresh page
- [ ] Next request uses a **new** `client_id`; localStorage entry is replaced with a fresh `createdAt`

4. Storage-disabled fallback (private context / storage blocked)

- [ ] Open the page in Safari private tab (or Chrome with cookies/storage blocked)
- [ ] Requests still carry an `X-Client-Id` / `?client_id=` value
- [ ] Value stays stable while the tab is open, changes after closing/reopening

# Background

Native `EventSource` does not support custom request headers per the WHATWG spec. Adding a polyfill (`event-source-polyfill` etc.) would let us mirror the POST header on SSE but adds a runtime dependency and a known set of reconnect/readyState bugs that we do not want to trade for pure cosmetic consistency. Query string is equally opaque and equally protected by TLS; the backend already accepts both.

The 1-day TTL bounds the long-term correlation window of a browser to a stable session key and gives orphan sessions on the backend an upper lifetime. Rotation happens only on the next call after a new tab opens or an existing one is refreshed, so a long-running tab keeps the same `client_id` for the whole session.

## Summary by CodeRabbit

* **New Features**
  * Balance watcher requests now include an `X-Client-Id` header, and the balances SSE stream now includes a `client_id` query parameter.
  * Added a stable, browser-persisted client identifier reused for up to one day.
* **Bug Fixes**
  * When browser storage is unavailable, the identifier falls back to a consistent in-tab value.
  * Legacy/unexpected stored values are ignored and a new identifier is generated.
* **Tests**
  * Added/updated coverage for identifier persistence, expiration, storage-failure fallback, and inclusion in headers/query parameters.

### CODE REVIEW

**Review by elena-zh — APPROVED** [review]
Works as described!

File: libs/balances-and-allowances/src/balancesWatcher/clientId.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+// Rotated only on the next call after a new tab opens or an existing one is refreshed.
+const CLIENT_ID_TTL_MS = ms`1 day`
+
+interface StoredClientId {
+  id: string
+  createdAt: number
+}
+
+let inMemoryFallback: StoredClientId | null = null
```
Comment (Danziger) [inline libs/balances-and-allowances/src/balancesWatcher/clientId.ts]: I'd probably use this as a cache to avoid constantly reading from localStorage. The logic in this function could look something like this:

- Check if there's a non-expired cached value and return it.
- If there isn't or it is expired, read it from localStorage. If there isn't any or if it's also expired, generate a new one.
- Cache it in memory.
- Persist it in localStorage. Wrap only this part in try..catch and log the error.

If localStorage value is changed on a different tab (e.g. because it expired), the next one to run will also notice the cached token is expired and it will load the fresh value from localStorage. In any case, you can also listen for storage events if you want to sync it immediately.

**Review by fairlighteth — COMMENTED** [review]
<details>
<summary>⚠️ <strong>AI Review</strong> (Codex GPT-5, worked 4m): client_id can rotate between POST and SSE</summary>

**Finding:** [BLOCKING] Pin the balances-watcher `client_id` across the POST → SSE handshake

- Location: `libs/balances-and-allowances/src/hooks/balancesWatcherSessionController.ts:128`
- This one is important: the PR says the backend indexes sessions by `(chainId, owner, clientId)`, and the checklist requires the POST `X-Client-Id` and SSE `client_id` to be the same UUID.
- Current code calls `getBalancesWatcherClientId()` independently in `createBalancesWatcherSession` and `subscribeToBalancesEvents`. If the stored ID is still fresh when the POST starts but crosses the 24h TTL before the POST resolves and `openStream()` runs, the POST registers the old ID while the SSE URL rotates to a new ID.
- Result: the SSE stream can look up a different backend session than the one just created.

**Suggested fix**

- Generate/read the client ID once per session attempt and pass that same value into both helpers, or make `getBalancesWatcherClientId()` cache the active tab ID so TTL is only re-evaluated on a new page load.
- Add a regression test where the stored ID expires between `createBalancesWatcherSession` and `subscribeToBalancesEvents`, and assert both requests still use the same ID.

<details>
<summary>Review scope and related context</summary>

- Existing review context did not cover this: CodeRabbit reported no actionable comments, and there are no review threads.
- The existing tests cover fresh storage, expired storage, POST header, and SSE query parameter separately, but not the controller boundary where the same session attempt must keep one ID across both requests.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify this finding against current code. Fix only if still valid, keep the change minimal, and validate with targeted tests.

Context:
- createSessionController calls createBalancesWatcherSession(), waits for it to resolve, then opens subscribeToBalancesEvents().
- createBalancesWatcherSession and subscribeToBalancesEvents each call getBalancesWatcherClientId() separately.
- If the 24h TTL boundary is crossed between those calls, POST and SSE can use different client IDs.
- Expected fix: pin one client ID per session attempt or cache the active tab ID until reload, then add a regression test for the TTL-boundary handshake.
```

</details>

_Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

File: libs/balances-and-allowances/src/balancesWatcher/clientId.ts (hunk 1 above)
Comment (limitofzero) [reply libs/balances-and-allowances/src/balancesWatcher/clientId.ts]: it's not a strong guard, it's just needed to update session id time-to-time. We actually don't need to update it at all in most cases, we need just a rotation here. In my opinion: caching it in memory + immediate sync overcomplicates it

@alfetopito I need your opinion here

Comment (Danziger) [reply libs/balances-and-allowances/src/balancesWatcher/clientId.ts]: My concern is that reading from `localStorage` is relatively slow and blocking, so while I doubt we call this function too often, I'd rather avoid it. And the in-memory caching change might just a matter of re-arranging the logic you already have with `inMemoryFallback` anyway:

```
let cachedClientId: StoredClientId | null = null // This replaces inMemoryFallback

export function getBalancesWatcherClientId(): string {
  if (cachedClientId) return cachedClientId.id

  const stored = parseStored(localStorage.getItem(STORAGE_KEY))

  if (stored && !isExpired(stored)) {
    cachedClientId = stored
  } else {
    cachedClientId = { id: crypto.randomUUID(), createdAt: Date.now() }

    try {
      localStorage.setItem(STORAGE_KEY, JSON.stringify(cachedClientId))
    } catch (err) {
      console.warn("Error storing StoredClientId:", err)
    }
  }

  return cachedClientId.id
}
```

Synching with storage events might not be needed given that we check the expiration anyway, I'm just wondering if there might be a small race condition between tabs (and if that's relevant at all or not), where one tab notices the token is expired, updates it, and another tab still sends a request with the old token in some very rare case.

Comment (alfetopito) [reply libs/balances-and-allowances/src/balancesWatcher/clientId.ts]: Partially agree with both.

I don't think we should over complicate it, the current implementation is only the id for fallbacks. We shouldn't need to read from localStorage that often.

We should however improve the try/catch to not throw away the generated id in case the setItem fails.
But that's also a minor thing IMO.

Comment (limitofzero) [reply libs/balances-and-allowances/src/balancesWatcher/clientId.ts]: saved it to variable to not read from localStorage every time

**Comment by limitofzero** [comment]
> ⚠️ **AI Review** (Codex GPT-5, worked 4m): client_id can rotate between POST and SSE

done

**Review by Danziger — APPROVED** [review]

**Comment by fairlighteth** [comment]
<details>
<summary>✅ <strong>AI Review</strong> (Codex GPT-5, worked 4m): follow-up addressed</summary>

**Rechecked**

- Author response: [marked as done](https://github.com/cowprotocol/cowswap/pull/7823#issuecomment-4935812281).
- Code path: `libs/balances-and-allowances/src/balancesWatcher/clientId.ts:18`.
- Test coverage: `clientId.test.ts:90` now crosses the TTL boundary and asserts the ID remains stable. The [balances-and-allowances CI suite](https://github.com/cowprotocol/cowswap/actions/runs/29105058271/job/86403604005) passed all 8 suites and 72 tests.

**Result:** Fixed. The first call now caches the selected ID for the tab lifetime, so the later SSE call cannot rotate away from the ID used by the POST. Rotation occurs only after a page/module reload, matching the updated PR contract. No new non-duplicate findings surfaced.

<details>
<summary>🤖 Verification notes for AI agents</summary>

```text
Verify the prior finding against current code only.

Prior finding:
- POST and SSE called getBalancesWatcherClientId() separately, allowing TTL rotation between them.

Current verification:
- A module-scoped cachedClientId is populated on the first call and returned before subsequent storage/TTL checks.
- The regression test advances time beyond one day and confirms both calls return the same ID.
- The balances-and-allowances CI suite passed.
```

</details>

_Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>
